"Connection reset by peer" during file upload

Akka server randomly fails with connection rest by peer during uploads if response is sent before all chunks are processed. For example when entity size limit is reached. Similar problems can be observed if upload is done through browser, sometimes response is sent, sometimes browser waits for full upload, sometimes connection is closed. I think I managed to isolate the problem and there is code to do that:  

```
[debug]   Classpath:
[debug] 	/tmp/sbt_c914386c/job-1/target/f6620d8f/con_test_2.12-0.1-SNAPSHOT.jar
[debug] 	/tmp/sbt_c914386c/target/f2e496f2/scala-library.jar
[debug] 	/tmp/sbt_c914386c/target/437cfe40/akka-actor_2.12-2.5.4.jar
[debug] 	/tmp/sbt_c914386c/target/2cf7a6cc/config-1.3.1.jar
[debug] 	/tmp/sbt_c914386c/target/1e6f1e74/scala-java8-compat_2.12-0.8.0.jar
[debug] 	/tmp/sbt_c914386c/target/395605a0/akka-http-spray-json_2.12-10.0.10.jar
[debug] 	/tmp/sbt_c914386c/target/4cbea03a/akka-http_2.12-10.0.10.jar
[debug] 	/tmp/sbt_c914386c/target/eb51e977/akka-http-core_2.12-10.0.10.jar
[debug] 	/tmp/sbt_c914386c/target/b6ba857b/akka-parsing_2.12-10.0.10.jar
[debug] 	/tmp/sbt_c914386c/target/af1a9f8f/akka-stream_2.12-2.5.4.jar
[debug] 	/tmp/sbt_c914386c/target/1b1c9116/reactive-streams-1.0.1.jar
[debug] 	/tmp/sbt_c914386c/target/7497c001/ssl-config-core_2.12-0.2.1.jar
[debug] 	/tmp/sbt_c914386c/target/7c5f25a2/scala-parser-combinators_2.12-1.0.4.jar
[debug] 	/tmp/sbt_c914386c/target/6e01a672/spray-json_2.12-1.3.3.jar
[debug] 	/tmp/sbt_c914386c/target/457bfc65/akka-http-xml_2.12-10.0.10.jar
[debug] 	/tmp/sbt_c914386c/target/e22de336/scala-xml_2.12-1.0.6.jar
[debug] 	/tmp/sbt_c914386c/target/b5d88662/akka-slf4j_2.12-2.5.4.jar
[debug] 	/tmp/sbt_c914386c/target/432be7c9/slf4j-api-1.7.23.jar
[debug] Waiting for thread run-main-0 to terminate.

```

Sample code to reproduce bug: 

```
import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.HttpEntity.{Chunk, Chunked}
import akka.http.scaladsl.model._
import akka.http.scaladsl.server.Directives._
import akka.http.scaladsl.server.{ExceptionHandler, Route}
import akka.stream.{ActorMaterializer, Materializer}
import akka.stream.scaladsl.{Sink, Source}

import scala.concurrent.Await
import scala.concurrent.duration._
import scala.util.{Failure, Random, Success}

object ConnectionResetByPeerTest extends scala.App{
  implicit val system = ActorSystem("test-system")
  implicit val materializer: Materializer = ActorMaterializer()
  implicit val executor = system.dispatcher

  def entityStreamSizeExceptionHandler = ExceptionHandler {
    case _: EntityStreamSizeException => complete(StatusCodes.RequestEntityTooLarge)
  }

  val uploadSizeLimit = 1024 * 1024
  def upload: Route = handleExceptions(entityStreamSizeExceptionHandler){
    (post & path("upload")){
      withSizeLimit(uploadSizeLimit){
        extractRequestContext { ctx =>
          complete(ctx.request.entity.dataBytes.runWith(Sink.ignore).map(_ => StatusCodes.OK))
        }
      }
    }
  }

  val bind = Await.ready(Http().bindAndHandle( upload, "0.0.0.0", 8080), 1 minute)

  val flow = Http().superPool[Unit]()

  val chunkCount = 20000
  (1 to 10000) foreach { i=>
    val chunk = Chunk(Random.alphanumeric.take(100000).mkString)
    val source = Source(List.fill(chunkCount)(chunk))
    val entity = Chunked(ContentTypes.`text/plain(UTF-8)`, source)

    println("===========================")
    println("Iteration: "+i)
    println("===========================")

    val request = HttpRequest(method = HttpMethods.POST, uri = "http://localhost:8080/upload", entity = entity)
    val responseFuture = Source.single(request -> ((): Unit)).via(flow).completionTimeout(1 minute).runWith(Sink.head)

    Await.ready(responseFuture.map(_._1), 1 minute).value.map(r => r.flatten) match {
      case Some(Success(result)) => result
      case Some(Failure(error)) => throw new AssertionError(error)
      case None => sys.error("Future reached timeout ")
    }

  }

  bind.foreach(_.unbind())
  println("DONE")
}
```
Error: 

```
[INFO] [09/25/2017 12:37:03.024] [test-system-akka.actor.default-dispatcher-23] [akka://test-system/system/IO-TCP/selectors/$a/19] Message [akka.io.Tcp$Write] from Actor[akka://test-system/user/StreamSupervisor-1/$$j#1726155251] to Actor[akka://test-system/system/IO-TCP/selectors/$a/19#229162384] was not delivered. [6] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
===========================
Iteration: 11
===========================
[INFO] [09/25/2017 12:37:03.053] [test-system-akka.actor.default-dispatcher-3] [akka://test-system/system/IO-TCP/selectors/$a/21] Message [akka.io.Tcp$Write] from Actor[akka://test-system/user/StreamSupervisor-1/$$k#-874021409] to Actor[akka://test-system/system/IO-TCP/selectors/$a/21#1225581899] was not delivered. [7] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
===========================
Iteration: 12
===========================
===========================
Iteration: 13
===========================
[error] (run-main-0) java.lang.AssertionError: akka.stream.StreamTcpException: The connection closed with error: Connection reset by peer
[error] java.lang.AssertionError: akka.stream.StreamTcpException: The connection closed with error: Connection reset by peer
[error] 	at ConnectionResetByPeerTest$.$anonfun$new$1(ConnectionResetByPeerTest.scala:54)
[error] 	at ConnectionResetByPeerTest$.$anonfun$new$1$adapted(ConnectionResetByPeerTest.scala:40)
[error] 	at scala.collection.immutable.Range.foreach(Range.scala:156)
[error] 	at ConnectionResetByPeerTest$.delayedEndpoint$ConnectionResetByPeerTest$1(ConnectionResetByPeerTest.scala:40)
[error] 	at ConnectionResetByPeerTest$delayedInit$body.apply(ConnectionResetByPeerTest.scala:15)
[error] 	at scala.Function0.apply$mcV$sp(Function0.scala:34)
[error] 	at scala.Function0.apply$mcV$sp$(Function0.scala:34)
[error] 	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
[error] 	at scala.App.$anonfun$main$1$adapted(App.scala:76)
[error] 	at scala.collection.immutable.List.foreach(List.scala:389)
[error] 	at scala.App.main(App.scala:76)
[error] 	at scala.App.main$(App.scala:74)
[error] 	at ConnectionResetByPeerTest$.main(ConnectionResetByPeerTest.scala:15)
[error] 	at ConnectionResetByPeerTest.main(ConnectionResetByPeerTest.scala)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error] 	at java.lang.reflect.Method.invoke(Method.java:498)
[error] 	at sbt.Run.invokeMain(Run.scala:89)
[error] 	at sbt.Run.run0(Run.scala:83)
[error] 	at sbt.Run.execute$1(Run.scala:61)
[error] 	at sbt.Run.$anonfun$run$4(Run.scala:73)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] 	at sbt.util.InterfaceUtil$$anon$1.get(InterfaceUtil.scala:10)
[error] 	at sbt.TrapExit$App.run(TrapExit.scala:252)
[error] 	at java.lang.Thread.run(Thread.java:748)
[error] Caused by: akka.stream.StreamTcpException: The connection closed with error: Connection reset by peer
[debug] 	Thread run-main-0 exited.
[debug] Waiting for thread test-system-akka.actor.default-blocking-io-dispatcher-10 to terminate.

```

With the new connection pool implementation this fails right away with: 

```
[ERROR] [04/03/2018 12:52:43.244] [test-system-akka.actor.default-dispatcher-11] [test-system/Pool(shared->http://localhost:8080)] [0 (WaitingForEndOfRequestEntity)] Slot execution failed. That's probably a bug. Please file a bug at https://github.com/akka/akka-http/issues. Slot is restarted.
java.lang.IllegalStateException: Cannot [onResponseReceived] when in state [WaitingForEndOfRequestEntity]
	at akka.http.impl.engine.client.pool.SlotState.illegalState(SlotState.scala:84)
	at akka.http.impl.engine.client.pool.SlotState.onResponseReceived(SlotState.scala:61)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Event$.$anonfun$onResponseReceived$1(NewHostConnectionPool.scala:130)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.runOneTransition$1(NewHostConnectionPool.scala:208)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.loop$1(NewHostConnectionPool.scala:282)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.updateState(NewHostConnectionPool.scala:291)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.onResponseReceived(NewHostConnectionPool.scala:182)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.$anonfun$onPush$2(NewHostConnectionPool.scala:406)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.$anonfun$onPush$2$adapted(NewHostConnectionPool.scala:406)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.withSlot(NewHostConnectionPool.scala:371)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.onPush(NewHostConnectionPool.scala:406)
	at akka.stream.stage.GraphStageLogic$SubSinkInlet.$anonfun$_sink$1(GraphStage.scala:1282)
	at akka.stream.stage.GraphStageLogic$SubSinkInlet.$anonfun$_sink$1$adapted(GraphStage.scala:1277)
	at akka.stream.impl.fusing.GraphInterpreter.runAsyncInput(GraphInterpreter.scala:454)
	at akka.stream.impl.fusing.GraphInterpreterShell$AsyncInput.execute(ActorGraphInterpreter.scala:465)
	at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:560)
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:742)
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$shortCircuitBatch(ActorGraphInterpreter.scala:732)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:758)
	at akka.actor.Actor.aroundReceive(Actor.scala:517)
	at akka.actor.Actor.aroundReceive$(Actor.scala:515)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:667)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:590)
	at akka.actor.ActorCell.invoke(ActorCell.scala:559)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

Swtiching to the legacy implementation doesn't help. The full error message is:

```
root [ERROR] [05/14/2018 16:02:33.496] [Server-akka.actor.default-dispatcher-46] [akka://Server/user/unhandled-messages] Unhandled message `Failure(akka.stream.StreamTcpException: The connection closed with er
ror: Connection reset by peer)` to `Actor[akka://Server/user/$Ab#788992179]` from `Actor[akka://Server/deadLetters]`.
```

The original issue can still be reproduced on current master. I think the server behavior is reasonable: If the entity size limit is hit, we sent an early response and reset the connection afterwards (no other way to fend off the extra unwanted incoming data).

Still the client side fails with an exception depending on timings in this case which should never happen:

```
[info] [ERROR] [03/23/2020 14:21:28.527] [test-system-akka.actor.default-dispatcher-5] [test-system/Pool(shared->http://localhost:8080)] [0 (Unconnected)] Slot execution failed. That's probably a bug. Please file a bug at https://github.com/akka/akka-http/issues. Slot is restarted.
[info] java.lang.IllegalStateException: Cannot open connection when slot still has an open connection
[info] 	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.openConnection(NewHostConnectionPool.scala:389)
[info] 	at akka.http.impl.engine.client.pool.SlotState$UnconnectedState.onNewRequest(SlotState.scala:182)
[info] 	at akka.http.impl.engine.client.pool.SlotState$UnconnectedState.onNewRequest$(SlotState.scala:181)
[info] 	at akka.http.impl.engine.client.pool.SlotState$Unconnected$.onNewRequest(SlotState.scala:192)
[info] 	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Event$.$anonfun$onNewRequest$1(NewHostConnectionPool.scala:154)
[info] 	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.runOneTransition$1(NewHostConnectionPool.scala:259)
[info] 	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.loop$1(NewHostConnectionPool.scala:341)
[info] 	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.updateState(NewHostConnectionPool.scala:350)
[info] 	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.onNewRequest(NewHostConnectionPool.scala:222)
[info] 	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1.dispatchRequest(NewHostConnectionPool.scala:122)
[info] 	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1.onPush(NewHostConnectionPool.scala:87)
[info] 	at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:523)
[info] 	at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:480)
[info] 	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:376)
[info] 	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:606)
[info] 	at akka.stream.impl.fusing.GraphInterpreterShell$AsyncInput.execute(ActorGraphInterpreter.scala:485)
[info] 	at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:581)
[info] 	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:749)
[info] 	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:764)
[info] 	at akka.actor.Actor.aroundReceive(Actor.scala:539)
[info] 	at akka.actor.Actor.aroundReceive$(Actor.scala:537)
[info] 	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:671)
[info] 	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:612)
[info] 	at akka.actor.ActorCell.invoke(ActorCell.scala:581)
[info] 	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:268)
[info] 	at akka.dispatch.Mailbox.run(Mailbox.scala:229)
[info] 	at akka.dispatch.Mailbox.exec(Mailbox.scala:241)
[info] 	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
[info] 	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
[info] 	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
[info] 	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala
@@ -257,6 +257,7 @@ private[client] object NewHostConnectionPool {
 
                 debug(s"Before event [${event.name}] In state [${state.name}] for [${timeInState / 1000000} ms]")
                 state = event.transition(state, this, arg)
+                require(state != Unconnected, "Slot must not change to Unconnected state") // Use ToBeClosed or Failed instead from state impls
                 debug(s"After event [${event.name}] State change [${previousState.name}] -> [${state.name}]")
 
                 state.stateTimeout match {

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/SlotState.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/SlotState.scala
@@ -340,12 +340,12 @@ private[pool] object SlotState {
     final override def isIdle = false
 
     override def onRequestEntityCompleted(ctx: SlotContext): SlotState =
-      if (ctx.isConnectionClosed) Unconnected
+      if (ctx.isConnectionClosed) ToBeClosed
       else Idle
     override def onRequestEntityFailed(ctx: SlotContext, cause: Throwable): SlotState =
-      if (ctx.isConnectionClosed) Unconnected
+      if (ctx.isConnectionClosed) ToBeClosed // ignore error here
       else Idle
-    override def onConnectionCompleted(ctx: SlotContext): SlotState = Unconnected
-    override def onConnectionFailed(ctx: SlotContext, cause: Throwable): SlotState = Unconnected
+    override def onConnectionCompleted(ctx: SlotContext): SlotState = ToBeClosed
+    override def onConnectionFailed(ctx: SlotContext, cause: Throwable): SlotState = Failed(cause)
   }
 }

--- a/akka-http-core/src/test/scala/akka/http/impl/engine/client/HostConnectionPoolSpec.scala
+++ b/akka-http-core/src/test/scala/akka/http/impl/engine/client/HostConnectionPoolSpec.scala
@@ -137,6 +137,31 @@ class HostConnectionPoolSpec extends AkkaSpecWithMaterializer(
         resBodyIn.request(1) // FIXME: should we support eager completion here? (reason is substreamHandler in PrepareResponse)
         resBodyIn.expectComplete()
       }
+      "not crash slot when connection is aborted after an early response while waiting for rest of request entity" inWithShutdown new SetupWithServerProbes {
+        // #1439
+        val reqEntityProbe = pushChunkedRequest()
+        responseOut.request(1)
+
+        val conn1 = expectNextConnection()
+        val reqInProbe = conn1.expectChunkedRequestBytesAsProbe()
+
+        // request coming in
+        reqEntityProbe.sendNext(ByteString("test"))
+        reqInProbe.expectUtf8EncodedString("test")
+
+        // ...
+        reqEntityProbe.sendNext(ByteString("test2"))
+        reqInProbe.expectUtf8EncodedString("test2")
+
+        conn1.failHandler(new RuntimeException("test"))
+
+        // In #1439, the next request now ran into a broken slot and the request is dropped to the floor
+        pushRequest()
+        val conn2 = expectNextConnection()
+        conn2.expectRequest()
+        conn2.pushResponse()
+        expectResponse()
+      }
       "open up to max-connections when enough requests are pending" inWithShutdown new SetupWithServerProbes(_.withMaxConnections(2)) {
         pushRequest(HttpRequest(uri = "/1"))
         val conn1 = expectNextConnection()

